content/en/blog/_posts/kubernetes-1-27-efficient-selinux-relabeling-beta.md

```diff
@@ -0,0 +1,126 @@
+---
+
+**Author:** Jan Šafránek (Red Hat)
+
+# The problem
+
+On Linux with Security-Enhanced Linux (SELinux) enabled, it's traditionally
+the container runtime who applies SELinux labels to a Pod and all it's volumes.
+Kubernetes only provides the SELinux label from Pod's Security Context fields
+to the container runtime.
+
+The container runtime then recursively changes SELinux label on all files that
+are visible to the Pod's containers. This can be time-consuming if there are
+many files on the volume, especially when the volume is on a remote filesystem.
+
+{{< note >}}
+If a container uses `subPath` of a volume, only that `subPath` of the whole
+volume is relabeled. This allows two pods that have two different SELinux labels
+to use the same volume, as long as they use different subpaths of it.
+{{< /note >}}
+
+{{< note >}}
+If a Pod does not have any SELinux label assigned in Kubernetes API, the
+container runtime assigns a unique random one, so a process that potentially
+escapes the container boundary cannot access data of any other container on the
+host. The container runtime still recursively relabels all pod volumes with this
+random SELinux label.
+{{< /note >}}
+
+# Improvement using mount options
+
+If a Pod and its volume meet **all** of the following conditions, Kubernetes will
+_mount_ the volume directly with the right SELinux label. Such mount will happen
+in a constant time and the container runtime will not need to recursively
+relabel any files on it.
+
+1. The operating system must support SELinux.
+
+   Without SELinux support detected, kubelet and the container runtime does not
+   do anything with regard to SELinux.
+
+1. The [feature gates](/docs/reference/command-line-tools-reference/feature-gates/)
+   `ReadWriteOncePod` and `SELinuxMountReadWriteOncePod` must be enabled.
+   These feature gates are Beta in Kubernetes 1.27 and Alpha in 1.25.
+
+   With any of these feature gates disabled, SELinux labels will be always
+   applied by the container runtime by a recursive walk through the volume
+   (or its subPaths).
+
+1. The Pod must have at least `seLinuxOptions.level` assigned in its [Pod Security Context](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#security-context) or all Pod containers must have it set in their [Security Contexts](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#security-context-1).
+   Kubernetes will read the default `user`, `role` and `type` from the operating
+   system defaults (typically `system_u`, `system_r` and `container_t`).
+
+   Without Kubernetes knowing at least the SELinux `level`, the container
+   runtime will assign a random one _after_ the volumes are mounted. The
+   container runtime will still relabel the volumes recursively in that case.
+
+1. The volume must be a Persistent Volume with
+   [Access Mode](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes)
```

Comment:
```sugggestion
   [Access Mode](/docs/concepts/storage/persistent-volumes/#access-modes).
```

Reply:
I removed `https://kubernetes.io` from all links to existing docs.